### PR TITLE
manual: document ways of obtaining source hashes

### PR DESCRIPTION
**Pull request by danbst**
... and security nuances.

Related: https://github.com/NixOS/nixpkgs/pull/53754

cc @7c6f434c @grahamc

### CODE REVIEW

**Comment by samueldr** [comment]
Rendered with state as of 663b8cc9298c87940dde596e292006fad02393c1

image

File: doc/coding-conventions.xml [hunk 1]
```diff
@@ -876,6 +876,107 @@ src = fetchFromGitHub {
...
+   <listitem>
+    <para>
+     Prefetch by package source (with <literal>nix-prefetch-url
+     '&lt;nixpkgs&gt;' -A <replaceable>PACKAGE</replaceable>.src</literal>,
+     where <replaceable>PACKAGE</replaceable> is package attribute name). Hash
+     is printed to stdout.
+    </para>
+    <para>
+     This works well when you've upgraded existing package version and want to
+     find out new hash, but is useless if package doesn't have top-level
```
Comment (7c6f434c) [inline doc/coding-conventions.xml]: It's not so much a question of top-level attribute, as of being accessible via an attribute path.

File: doc/coding-conventions.xml [hunk 2]
```diff
@@ -876,6 +876,107 @@ src = fetchFromGitHub {
...
+     <link xlink:href="https://github.com/NixOS/nixpkgs/blob/d2ab091dd308b99e4912b805a5eb088dd536adb9/pkgs/applications/video/kodi/default.nix#L73">
+     one of <literal>kodi</literal> dependencies</link>). The easiest way then
+     would be replace hash with a fake one and rebuild. Nix build will fail and
+     error message will contain wanted hash.
+    </para>
+   </listitem>
+  </itemizedlist>
+
+  <section xml:id="sec-source-hashes-security">
+   <title>Obtaining hashes securely</title>
```
Comment (7c6f434c) [inline doc/coding-conventions.xml]: Maybe note that the threat model discussed here is MITM close to the developer's network?

Whatever you do, MITM close to the server side (or a succesful server intruder) can even get a «valid» certificate via DV…

**Comment by 7c6f434c** [comment]
Technically speaking, there is a one more crazy-ish way of obtaining hashes that I sometimes use: fake hashes + Ctrl-c + copy the URL from the progress output + `nix-prefetch-…`

This way I do get the exact URL for free, but TLS is checked and the file is only downloaded once.

**Comment by 7c6f434c** [comment]
By the way, does git verify correctness of revision hashes by default? Does `fetchgit` enable/disable this? Fetching a known-good commit hash with no certificate validation might be or not be safe, and in both cases it could be useful to mention.

File: doc/coding-conventions.xml [hunk 3]
```diff
@@ -876,6 +876,107 @@ src = fetchFromGitHub {
...
+     Upstream provided hash: use it when upstream provides
+     <literal>sha256</literal> or <literal>sha512</literal> (when upstream
+     provides <literal>md5</literal>, don't use it, compute
+     <literal>sha256</literal> instead).
+    </para>
+    <para>
+     A little nuance is that <literal>nix-prefetch-*</literal> tools produce
+     hash encoded with <literal>base32</literal>, but upstream usually provides
+     hexadecimal (<literal>base16</literal>) encoding. Fetchers understand both
+     formats. Nixpkgs doesn't stadartize on any one format.
```
Comment (Mic92) [inline doc/coding-conventions.xml]:
```suggestion
     formats. Nixpkgs does not standardize on any one format.
```

**Comment by danbst** [comment]
@Mic92 applied!

@7c6f434c rephrased the security section. Rendered

screenshot from 2019-01-17 12-31-28

**Comment by Mic92** [comment]
Thanks!

**Pull request merged**